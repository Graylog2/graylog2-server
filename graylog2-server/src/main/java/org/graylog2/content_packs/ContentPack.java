/**
 * This file is part of Graylog.
 *
 * Graylog is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Graylog is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.graylog2.content_packs;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import org.mongojack.Id;
import org.mongojack.ObjectId;

import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import java.util.Collections;
import java.util.Set;

@JsonAutoDetect
@JsonInclude(JsonInclude.Include.NON_NULL)
public class ContentPack {
    @Id
    @ObjectId
    @JsonProperty
    @NotBlank
    private String id;
    @JsonProperty
    @NotBlank
    private String name;
    @JsonProperty
    @NotBlank
    private String version;
    @JsonProperty
    private String description;
    @JsonProperty
    private String summary;
    @JsonProperty
    @NotNull
    private Set<String> states = Collections.emptySet();

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getVersion() {
        return version;
    }

    public void setVersion(String version) {
        this.version = version;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getSummary() {
        return summary;
    }

    public void setSummary(String summary) {
        this.summary = summary;
    }

    public Set<String> getStates() {
        return states;
    }

    public void setStates(Set<String> states) {
        this.states = states;
    }
}
