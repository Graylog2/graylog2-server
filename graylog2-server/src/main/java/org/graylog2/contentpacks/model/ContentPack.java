/**
 * This file is part of Graylog.
 *
 * Graylog is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Graylog is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.graylog2.contentpacks.model;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import org.graylog2.contentpacks.model.constraints.Constraint;

import java.util.Set;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.EXISTING_PROPERTY, property = Versioned.FIELD_META_VERSION, defaultImpl = LegacyContentPack.class)
@JsonSubTypes({
        @JsonSubTypes.Type(value = LegacyContentPack.class),
        @JsonSubTypes.Type(value = ContentPackV1.class, name = ContentPackV1.VERSION)
})
public interface ContentPack extends Identified, Revisioned, Versioned {
    interface ContentPackBuilder<SELF> extends IdBuilder<SELF>, RevisionBuilder<SELF>, VersionBuilder<SELF> {
    }
}
