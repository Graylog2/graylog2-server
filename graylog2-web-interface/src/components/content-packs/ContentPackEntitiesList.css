:local(.scrollable) tbody {
    max-height: 300px;
    overflow: hidden;
    overflow-y: auto;
    display:block;
}

:local(.scrollable) thead, :local(.scrollable) tbody tr {
    display:table;
    width:100%;
    table-layout:fixed;
}

:local(.bigColumns) {
    word-wrap: break-word;
}
