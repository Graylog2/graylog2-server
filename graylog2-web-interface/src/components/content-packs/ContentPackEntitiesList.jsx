import PropTypes from 'prop-types';
import React from 'react';

import { Button, Modal, ButtonToolbar } from 'react-bootstrap';
import { SearchForm, DataTable } from 'components/common';
import BootstrapModalWrapper from 'components/bootstrap/BootstrapModalWrapper';

import ContentPackApplyParameter from './ContentPackApplyParameter';
import ContentPackEntityConfig from './ContentPackEntityConfig';

import ContentPackEntitiesListStyle from './ContentPackEntitiesList.css';

class ContentPackEntitiesList extends React.Component {
  static propTypes = {
    contentPack: PropTypes.object.isRequired,
    appliedParameter: PropTypes.object,
    onParameterApply: PropTypes.func,
    onParameterClear: PropTypes.func,
    readOnly: PropTypes.bool,
  };

  static defaultProps = {
    appliedParameter: {},
    onParameterClear: () => {},
    onParameterApply: () => {},
    readOnly: false,
  };

  constructor(props) {
    super(props);
    this.state = {
      filteredEntities: props.contentPack.entities || [],
      filter: undefined,
    };
  }

  componentWillReceiveProps(newProps) {
    this._filterEntities(this.state.filter, newProps.contentPack.entities);
  }

  _filterEntities = (filter, entitiesArg) => {
    const entities = entitiesArg || this.props.contentPack.entities;
    if (!filter || filter.length <= 0) {
      this.setState({ filteredEntities: entities, filter: undefined });
      return;
    }

    const regexp = RegExp(filter, 'i');
    const filteredEntities = entities.filter((entity) => {
      return regexp.test(entity.title) || regexp.test(entity.description);
    });
    this.setState({ filteredEntities: filteredEntities, filter: filter });
  };

  _entityIcon = (entity) => {
    if (!entity.fromServer) {
      return <span><i title="Content Pack" className={`fa fa-archive ${ContentPackEntitiesListStyle.contentPackEntity}`} /></span>;
    }
    return <span><i title="Server" className="fa fa-server" /></span>;
  };

  _entityRowFormatter = (entity) => {
    let applyModalRef;
    const applyParamComponent = (<ContentPackApplyParameter
      parameters={this.props.contentPack.parameters}
      entity={entity}
      appliedParameter={this.props.appliedParameter[entity.id]}
      onParameterApply={(key, value) => { this.props.onParameterApply(entity.id, key, value); }}
      onParameterClear={(key) => { this.props.onParameterClear(entity.id, key); }}
    />);

    const closeModal = () => {
      applyModalRef.close();
    };

    const open = () => {
      applyModalRef.open();
    };

    const applyModal = (
      <BootstrapModalWrapper ref={(node) => { applyModalRef = node; }} bsSize="large">
        <Modal.Header closeButton>
          <Modal.Title>Edit</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          {applyParamComponent}
        </Modal.Body>
        <Modal.Footer>
          <Button onClick={closeModal}>Close</Button>
        </Modal.Footer>
      </BootstrapModalWrapper>
    );

    let showModalRef;
    const entityComponent = (<ContentPackEntityConfig
      appliedParameter={this.props.appliedParameter[entity.id]}
      parameters={this.props.contentPack.parameters}
      entity={entity}
    />);

    const closeShowModal = () => {
      showModalRef.close();
    };

    const openShowModal = () => {
      showModalRef.open();
    };

    const showModal = (
      <BootstrapModalWrapper ref={(node) => { showModalRef = node; }} bsSize="large">
        <Modal.Header closeButton>
          <Modal.Title>Entity Config</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          {entityComponent}
        </Modal.Body>
        <Modal.Footer>
          <Button onClick={closeShowModal}>Close</Button>
        </Modal.Footer>
      </BootstrapModalWrapper>
    );

    const disableBtn = this.props.contentPack.parameters.length <= 0;
    const appliedParameterCount = (this.props.appliedParameter[entity.id] || []).length;
    return (
      <tr key={entity.id}>
        <td className={ContentPackEntitiesListStyle.bigColumns}>{entity.title}</td>
        <td>{entity.type.name}</td>
        <td className={ContentPackEntitiesListStyle.bigColumns}>{entity.description}</td>
        {!this.props.readOnly && <td>{this._entityIcon(entity)}</td>}
        {!this.props.readOnly && <td>{appliedParameterCount}</td>}
        <td>
          <ButtonToolbar>
            {!this.props.readOnly &&
            <Button bsStyle="primary"
                    bsSize="xs"
                    disabled={disableBtn}
                    onClick={() => {
                      open();
                    }}>
              Edit
            </Button>
            }
            <Button bsStyle="info"
                    bsSize="xs"
                    onClick={() => { openShowModal(); }}>
              Show
            </Button>
          </ButtonToolbar>
        </td>
        {!this.props.readOnly && applyModal}
        {showModal}
      </tr>
    );
  };

  render() {
    const headers = this.props.readOnly ?
      ['Title', 'Type', 'Description', 'Action'] :
      ['Title', 'Type', 'Description', 'Origin', 'Used Parameters', 'Action'];

    return (
      <div>
        <h2>Entity list</h2>
        <br />
        <SearchForm
          searchButtonLabel="Filter"
          onSearch={this._filterEntities}
          onReset={() => { this._filterEntities(''); }}
        />
        <DataTable
          id="entity-list"
          headers={headers}
          className={ContentPackEntitiesListStyle.scrollable}
          sortBy={entity => entity.type.name}
          filterKeys={[]}
          rows={this.state.filteredEntities}
          dataRowFormatter={this._entityRowFormatter}
        />
      </div>
    );
  }
}

export default ContentPackEntitiesList;
