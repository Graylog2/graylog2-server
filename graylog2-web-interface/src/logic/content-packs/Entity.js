import { Map } from 'immutable';
import { findIndex } from 'lodash';
import ValueRefHelper from 'util/ValueRefHelper';
import Constraint from './Constraint';

export default class Entity {
  constructor(v, type, id, data, fromServer = false, constraintValues = [], parameters = []) {
    const constraints = constraintValues.map((c) => {
      if (c instanceof Constraint) {
        return c;
      }
      return Constraint.fromJSON(c);
    });

    this._value = {
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
      parameters,
    };
  }

  static fromJSON(value, fromServer = true, parameters = []) {
    const { v, type, id, data, constraints } = value;
    return new Entity(v, type, id, data, fromServer, constraints, parameters);
  }

  get v() {
    return this._value.v;
  }

  get type() {
    return this._value.type;
  }

  get id() {
    return this._value.id;
  }

  get data() {
    return this._value.data;
  }

  get fromServer() {
    return this._value.fromServer;
  }

  get constraints() {
    return this._value.constraints;
  }

  get title() {
    let value = this.getValueFromData('title');
    if (!value) {
      value = this.getValueFromData('name');
    }
    return value || '';
  }

  get description() {
    return this.getValueFromData('description') || '';
  }

  /* eslint-disable-next-line class-methods-use-this */
  get isEntity() {
    return true;
  }

  /* implement custom instanceof */
  static [Symbol.hasInstance](obj) {
    if (obj.isEntity) {
      return true;
    }
    return false;
  }

  getValueFromData(key) {
    const { data } = this._value;
    if (!data || !data[key]) {
      return undefined;
    }

    if (ValueRefHelper.dataIsValueRef(data[key])) {
      const value = (data[key] || {})[ValueRefHelper.VALUE_REF_VALUE_FIELD];
      if (ValueRefHelper.dataValueIsParameter(data[key])) {
        const index = findIndex(this._value.parameters, { name: value });
        if (index >= 0 && this._value.parameters[index].default_value) {
          return this._value.parameters[index].default_value;
        }
      }
      return value;
    }
    return data[key];
  }

  toBuilder() {
    const {
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
      parameters,
    } = this._value;
    /* eslint-disable-next-line no-use-before-define */
    return new Builder(Map({
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
      parameters,
    }));
  }


  static builder() {
    /* eslint-disable-next-line no-use-before-define */
    return new Builder();
  }

  toJSON() {
    const {
      v,
      type,
      id,
      data,
      constraints,
    } = this._value;
    return {
      v,
      type,
      id,
      data,
      constraints,
    };
  }
}

class Builder {
  constructor(value = Map()) {
    this.value = value;
  }

  v(value) {
    this.value = this.value.set('v', value);
    return this;
  }

  type(value) {
    this.value = this.value.set('type', value);
    return this;
  }

  id(value) {
    this.value = this.value.set('id', value);
    return this;
  }

  data(value) {
    this.value = this.value.set('data', value);
    return this;
  }

  fromServer(value) {
    this.value = this.value.set('fromServer', value);
    return this;
  }

  constraints(value) {
    this.value = this.value.set('constraints', value);
    return this;
  }

  parameters(value) {
    this.value = this.value.set('parameters', value);
    return this;
  }

  build() {
    const {
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
      parameters,
    } = this.value.toObject();
    return new Entity(v, type, id, data, fromServer, constraints, parameters);
  }
}
