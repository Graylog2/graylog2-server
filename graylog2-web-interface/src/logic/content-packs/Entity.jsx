import { Map } from 'immutable';
import ValueRefHelper from 'util/ValueRefHelper';
import Constraint from './Constraint';

export default class Entity {
  constructor(v, type, id, data, fromServer = false, constraintValues = []) {
    const constraints = constraintValues.map((c) => {
      if (c instanceof Constraint) {
        return c;
      }
      return Constraint.fromJSON(c);
    });

    this._value = {
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
    };
  }

  static fromJSON(value, fromServer = true) {
    const { v, type, id, data, constraints } = value;
    return new Entity(v, type, id, data, fromServer, constraints);
  }

  get v() {
    return this._value.v;
  }

  get type() {
    return this._value.type;
  }

  get id() {
    return this._value.id;
  }

  get data() {
    return this._value.data;
  }

  get fromServer() {
    return this._value.fromServer;
  }

  get constraints() {
    return this._value.constraints;
  }

  get title() {
    let value = this.getValueFromData('title');
    if (!value) {
      value = this.getValueFromData('name');
    }
    return value || '';
  }

  get description() {
    return this.getValueFromData('description') || '';
  }

  getValueFromData(key) {
    const { data } = this._value;
    if (!data || !data[key]) {
      return undefined;
    }

    if (ValueRefHelper.dataIsValueRef(data[key])) {
      return (data[key] || {})[ValueRefHelper.VALUE_REF_VALUE_FIELD];
    }
    return data[key];
  }

  toBuilder() {
    const {
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
    } = this._value;
    /* eslint-disable-next-line no-use-before-define */
    return new Builder(Map({
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
    }));
  }


  static builder() {
    /* eslint-disable-next-line no-use-before-define */
    return new Builder();
  }

  toJSON() {
    const {
      v,
      type,
      id,
      data,
      constraints,
    } = this._value;
    return {
      v,
      type,
      id,
      data,
      constraints,
    };
  }
}

class Builder {
  constructor(value = Map()) {
    this.value = value;
  }

  v(value) {
    this.value = this.value.set('v', value);
    return this;
  }

  type(value) {
    this.value = this.value.set('type', value);
    return this;
  }

  id(value) {
    this.value = this.value.set('id', value);
    return this;
  }

  data(value) {
    this.value = this.value.set('data', value);
    return this;
  }

  fromServer(value) {
    this.value = this.value.set('fromServer', value);
    return this;
  }

  constraints(value) {
    this.value = this.value.set('constraints', value);
    return this;
  }

  build() {
    const {
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
    } = this.value.toObject();
    return new Entity(v, type, id, data, fromServer, constraints);
  }
}
