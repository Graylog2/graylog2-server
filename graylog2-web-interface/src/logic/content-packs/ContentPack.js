import { Map, Set } from 'immutable';
import { concat, remove } from 'lodash';
import uuid from 'uuid/v4';
import Entity from './Entity';

export default class ContentPack {
  constructor(v, id, rev, name, summary, description, vendor, url, parameters, entitieValues) {
    const entities = entitieValues.map((e) => {
      if (e instanceof Entity) {
        return e;
      }
      return Entity.fromJSON(e, false, parameters);
    });

    this._value = {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    };
  }

  get v() {
    return this._value.v;
  }

  get id() {
    return this._value.id;
  }

  get rev() {
    return this._value.rev;
  }

  get name() {
    return this._value.name;
  }

  get summary() {
    return this._value.summary;
  }

  get description() {
    return this._value.description;
  }

  get vendor() {
    return this._value.vendor;
  }

  get url() {
    return this._value.url;
  }

  get constraints() {
    return this._value.entities.reduce((acc, entity) => {
      return entity.constraints.reduce((result, constraint) => {
        return result.add(constraint);
      }, acc);
    }, Set());
  }

  get parameters() {
    return this._value.parameters;
  }

  get entities() {
    return this._value.entities;
  }

  toBuilder() {
    const {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    } = this._value;
    // eslint-disable-next-line no-use-before-define
    return new Builder(Map({
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    }));
  }

  toJSON() {
    const {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    } = this._value;

    const entitiesJSON = entities.map(e => e.toJSON());
    return {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities: entitiesJSON,
    };
  }

  static fromJSON(value) {
    const {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    } = value;
    return new ContentPack(
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    );
  }

  static builder() {
    // eslint-disable-next-line no-use-before-define
    return new Builder()
      .v(1)
      .id(uuid())
      .rev(1)
      .name('')
      .summary('')
      .description('')
      .vendor('')
      .url('')
      .parameters([])
      .entities([]);
  }
}

class Builder {
  constructor(value = Map()) {
    this.value = value;
  }

  v(value) {
    this.value = this.value.set('v', value);
    return this;
  }

  id(value) {
    this.value = this.value.set('id', value);
    return this;
  }

  rev(value) {
    this.value = this.value.set('rev', value);
    return this;
  }

  name(value) {
    this.value = this.value.set('name', value);
    return this;
  }

  summary(value) {
    this.value = this.value.set('summary', value);
    return this;
  }

  description(value) {
    this.value = this.value.set('description', value);
    return this;
  }

  vendor(value) {
    this.value = this.value.set('vendor', value);
    return this;
  }

  url(value) {
    this.value = this.value.set('url', value);
    return this;
  }

  parameters(value) {
    this.value = this.value.set('parameters', value);
    return this;
  }

  removeParameter(value) {
    const parameters = this.value.get('parameters').slice(0);
    remove(parameters, parameter => parameter.name === value.name);
    this.value = this.value.set('parameters', parameters);
    return this;
  }

  addParameter(value) {
    const parameters = this.value.get('parameters');
    const newParameters = concat(parameters, value);
    this.value = this.value.set('parameters', newParameters);
    return this;
  }

  entities(value) {
    this.value = this.value.set('entities', value);
    return this;
  }

  build() {
    const {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    } = this.value.toObject();
    return new ContentPack(
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    );
  }
}
