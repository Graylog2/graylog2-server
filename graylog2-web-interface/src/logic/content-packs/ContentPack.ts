/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import { Map, Set } from 'immutable';
import concat from 'lodash/concat';
import remove from 'lodash/remove';

import generateId from 'logic/generateId';
import type { ContentPackEntity, ContentPackParameter } from 'components/content-packs/Types';

import Entity from './Entity';

export default class ContentPack {
  private _value: {
    v: number;
    id: string;
    rev: number;
    name: string;
    summary: string;
    description: string;
    vendor: string;
    url: string;
    parameters: ContentPackParameter[];
    entities: ContentPackEntity[];
  };

  constructor(
    v: number,
    id: string,
    rev: number,
    name: string,
    summary: string,
    description: string,
    vendor: string,
    url: string,
    parameters: ContentPackParameter[],
    entitieValues: ContentPackEntity[],
  ) {
    const entities = entitieValues.map((e) => {
      if (e instanceof Entity) {
        return e;
      }

      return Entity.fromJSON(e, false, parameters);
    });

    this._value = {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    };
  }

  get v() {
    return this._value.v;
  }

  get id() {
    return this._value.id;
  }

  get rev() {
    return this._value.rev;
  }

  get name() {
    return this._value.name;
  }

  get summary() {
    return this._value.summary;
  }

  get description() {
    return this._value.description;
  }

  get vendor() {
    return this._value.vendor;
  }

  get url() {
    return this._value.url;
  }

  get constraints() {
    return this._value.entities.reduce(
      (acc, entity) => entity.constraints.reduce((result, constraint) => result.add(constraint), acc),
      Set(),
    );
  }

  get parameters() {
    return this._value.parameters;
  }

  get entities() {
    return this._value.entities;
  }

  toBuilder() {
    const { v, id, rev, name, summary, description, vendor, url, parameters, entities } = this._value;

    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new Builder(
      Map({
        v,
        id,
        rev,
        name,
        summary,
        description,
        vendor,
        url,
        parameters,
        entities,
      }),
    );
  }

  toJSON() {
    const { v, id, rev, name, summary, description, vendor, url, parameters, entities } = this._value;

    const entitiesJSON = entities.map((e) => e.toJSON());

    return {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities: entitiesJSON,
    };
  }

  static fromJSON(value) {
    const { v, id, rev, name, summary, description, vendor, url, parameters, entities } = value;

    return new ContentPack(v, id, rev, name, summary, description, vendor, url, parameters, entities);
  }

  static builder() {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new Builder()
      .v(1)
      .id(generateId())
      .rev(1)
      .name('')
      .summary('')
      .description('')
      .vendor('')
      .url('')
      .parameters([])
      .entities([]);
  }
}

class Builder {
  private value: Map<string, unknown>;

  constructor(value: Map<string, unknown> = Map()) {
    this.value = value;
  }

  v(value: number) {
    this.value = this.value.set('v', value);

    return this;
  }

  id(value: string) {
    this.value = this.value.set('id', value);

    return this;
  }

  rev(value: number) {
    this.value = this.value.set('rev', value);

    return this;
  }

  name(value: string) {
    this.value = this.value.set('name', value);

    return this;
  }

  summary(value: string) {
    this.value = this.value.set('summary', value);

    return this;
  }

  description(value: string) {
    this.value = this.value.set('description', value);

    return this;
  }

  vendor(value: string) {
    this.value = this.value.set('vendor', value);

    return this;
  }

  url(value: string) {
    this.value = this.value.set('url', value);

    return this;
  }

  parameters(value: ContentPackParameter[]) {
    this.value = this.value.set('parameters', value);

    return this;
  }

  private getParameters() {
    return this.value.get('parameters') as ContentPackParameter[];
  }

  removeParameter(value: ContentPackParameter) {
    const parameters = this.getParameters().slice(0);

    remove(parameters, (parameter) => parameter.name === value.name);
    this.value = this.value.set('parameters', parameters);

    return this;
  }

  addParameter(value: ContentPackParameter) {
    const parameters = this.getParameters();
    const newParameters = concat(parameters, value);

    this.value = this.value.set('parameters', newParameters);

    return this;
  }

  entities(value: Array<ContentPackEntity>) {
    this.value = this.value.set('entities', value);

    return this;
  }

  build() {
    const { v, id, rev, name, summary, description, vendor, url, parameters, entities } = this.value.toObject();

    // @ts-ignore
    return new ContentPack(v, id, rev, name, summary, description, vendor, url, parameters, entities);
  }
}
