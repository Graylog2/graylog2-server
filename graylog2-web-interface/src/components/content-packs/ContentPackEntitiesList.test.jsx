import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import 'helpers/mocking/react-dom_mock';

import ContentPack from 'logic/content-packs/ContentPack';
import ContentPackEntitiesList from 'components/content-packs/ContentPackEntitiesList';
import Entity from 'logic/content-packs/Entity';

describe('<ContentPackEntitiesList />', () => {
  const parameter = {
    title: 'A parameter name',
    name: 'TITLE',
    description: 'A parameter descriptions',
    type: 'string',
    default_value: 'test',
  };

  const entity1 = Entity.builder()
    .id('111-beef')
    .type({ name: 'Input', version: '1' })
    .v('1.0')
    .data({
      title: { '@type': 'parameter', '@value': 'TITLE' },
      configuration: {
        listen_address: { '@type': 'string', '@value': '1.2.3.4' },
        port: { '@type': 'integer', '@value': '23' },
      },
    })
    .parameters([parameter])
    .build();

  const entity2 = Entity.builder()
    .id('121-beef')
    .type({ name: 'Input', version: '1' })
    .v('1.0')
    .data({
      name: { '@type': 'string', '@value': 'BadInput' },
      title: { '@type': 'string', '@value': 'A bad input' },
      configuration: {
        listen_address: { '@type': 'string', '@value': '1.2.3.4' },
        port: { '@type': 'integer', '@value': '22' },
      },
    })
    .fromServer(true)
    .build();

  const contentPack = ContentPack.builder()
    .entities([entity1, entity2])
    .parameters([parameter])
    .build();

  it('should render with empty entities', () => {
    const emptyContentPack = { entities: [] };
    const wrapper = renderer.create(<ContentPackEntitiesList contentPack={emptyContentPack} readOnly />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render with entities and parameters without readOnly', () => {
    const appliedParameter = { '111-beef': [{ configKey: 'title', paramName: 'A parameter name' }] };

    const wrapper = renderer.create(<ContentPackEntitiesList contentPack={contentPack}
                                                             appliedParameter={appliedParameter} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should filter entities', () => {
    const wrapper = mount(<ContentPackEntitiesList contentPack={contentPack} />);
    expect(wrapper.find("td[children='test']").exists()).toBe(true);
    wrapper.find('input').simulate('change', { target: { value: 'Bad' } });
    wrapper.find('form').simulate('submit');
    expect(wrapper.find("td[children='test']").exists()).toBe(false);
    wrapper.find("button[children='Reset']").simulate('click');
    expect(wrapper.find("td[children='test']").exists()).toBe(true);
  });
});
