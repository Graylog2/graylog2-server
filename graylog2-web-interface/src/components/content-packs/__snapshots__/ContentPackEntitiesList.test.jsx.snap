// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContentPackEntitiesList /> should render with empty entities 1`] = `
<div>
  <h2>
    Entity list
  </h2>
  <br />
  <div
    className="search"
    style={
      Object {
        "marginTop": 15,
      }
    }
  >
    <form
      className="form-inline"
      onSubmit={[Function]}
    >
      <div
        className="form-group has-feedback"
      >
        <input
          autoComplete="off"
          className="query form-control"
          id="common-search-form-query-input"
          onChange={[Function]}
          placeholder="Enter search query..."
          spellCheck="false"
          style={
            Object {
              "width": "auto",
            }
          }
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group"
        style={
          Object {
            "marginLeft": 5,
          }
        }
      >
        <button
          className="submit-button btn btn-default"
          disabled={false}
          type="submit"
        >
          Filter
        </button>
      </div>
      <div
        className="form-group"
        style={
          Object {
            "marginLeft": 5,
          }
        }
      >
        <button
          className="reset-button btn btn-default"
          disabled={false}
          onClick={[Function]}
          type="reset"
        >
          Reset
        </button>
      </div>
    </form>
  </div>
  <div>
    <div
      className="row "
    >
      <div
        className="col-md-12"
      >
        <div
          className="data-table table-responsive"
          id="entity-list"
        >
          <p>
            No data available.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<ContentPackEntitiesList /> should render with entities and parameters without readOnly 1`] = `
<div>
  <h2>
    Entity list
  </h2>
  <br />
  <div
    className="search"
    style={
      Object {
        "marginTop": 15,
      }
    }
  >
    <form
      className="form-inline"
      onSubmit={[Function]}
    >
      <div
        className="form-group has-feedback"
      >
        <input
          autoComplete="off"
          className="query form-control"
          id="common-search-form-query-input"
          onChange={[Function]}
          placeholder="Enter search query..."
          spellCheck="false"
          style={
            Object {
              "width": "auto",
            }
          }
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group"
        style={
          Object {
            "marginLeft": 5,
          }
        }
      >
        <button
          className="submit-button btn btn-default"
          disabled={false}
          type="submit"
        >
          Filter
        </button>
      </div>
      <div
        className="form-group"
        style={
          Object {
            "marginLeft": 5,
          }
        }
      >
        <button
          className="reset-button btn btn-default"
          disabled={false}
          onClick={[Function]}
          type="reset"
        >
          Reset
        </button>
      </div>
    </form>
  </div>
  <div>
    <div
      className="row "
    >
      <div
        className="col-md-12"
      >
        <div
          className="data-table table-responsive"
          id="entity-list"
        >
          <table
            className="table scrollable"
          >
            <thead>
              <tr>
                <th>
                  Title
                </th>
                <th>
                  Type
                </th>
                <th>
                  Description
                </th>
                <th>
                  Origin
                </th>
                <th>
                  Used Parameters
                </th>
                <th>
                  Action
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td
                  className="bigColumns"
                >
                  test
                </td>
                <td>
                  Input
                </td>
                <td
                  className="bigColumns"
                >
                  
                </td>
                <td>
                  <span>
                    <i
                      className="fa fa-archive contentPackEntity"
                      title="Content Pack"
                    />
                  </span>
                </td>
                <td>
                  1
                </td>
                <td>
                  <div
                    className="btn-toolbar"
                    role="toolbar"
                  >
                    <button
                      className="btn btn-xs btn-primary"
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      Edit
                    </button>
                    <button
                      className="btn btn-xs btn-info"
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      Show
                    </button>
                  </div>
                </td>
              </tr>
              <tr>
                <td
                  className="bigColumns"
                >
                  A bad input
                </td>
                <td>
                  Input
                </td>
                <td
                  className="bigColumns"
                >
                  
                </td>
                <td>
                  <span>
                    <i
                      className="fa fa-server"
                      title="Server"
                    />
                  </span>
                </td>
                <td>
                  0
                </td>
                <td>
                  <div
                    className="btn-toolbar"
                    role="toolbar"
                  >
                    <button
                      className="btn btn-xs btn-primary"
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      Edit
                    </button>
                    <button
                      className="btn btn-xs btn-info"
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      Show
                    </button>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
